File: attendance-system/backend/src/AttendanceSystem.Domain/Entities/Clarification.cs
using AttendanceSystem.Domain.Common;
using AttendanceSystem.Domain.Enums;

namespace AttendanceSystem.Domain.Entities;

/// <summary>
/// Represents a clarification request sent to an employee
/// </summary>
public class Clarification : BaseEntity
{
    public Guid EmployeeId { get; private set; }
    public Employee Employee { get; private set; } = null!;

    public Guid RequestedById { get; private set; }
    public string RequestedByName { get; private set; } = string.Empty;

    public DateOnly StartDate { get; private set; }
    public DateOnly EndDate { get; private set; }
    public string Question { get; private set; } = string.Empty;
    public ClarificationStatus Status { get; private set; } = ClarificationStatus.Pending;

    // Response
    public string? Response { get; private set; }
    public DateTime? RespondedAt { get; private set; }

    // Due date
    public DateTime? DueDate { get; private set; }

    // Attachments
    private readonly List<ClarificationAttachment> _attachments = new();
    public IReadOnlyCollection<ClarificationAttachment> Attachments => _attachments.AsReadOnly();

    private Clarification() { }

    public static Clarification Create(
        Guid employeeId,
        Guid requestedById,
        string requestedByName,
        DateOnly startDate,
        DateOnly endDate,
        string question,
        DateTime? dueDate = null)
    {
        var clarification = new Clarification
        {
            EmployeeId = employeeId,
            RequestedById = requestedById,
            RequestedByName = requestedByName,
            StartDate = startDate,
            EndDate = endDate,
            Question = question,
            DueDate = dueDate,
            Status = ClarificationStatus.Pending
        };

        clarification.AddDomainEvent(new ClarificationRequestedEvent(
            clarification.Id, employeeId, requestedById, startDate, endDate));

        return clarification;
    }

    public void Respond(string response)
    {
        if (Status != ClarificationStatus.Pending)
            throw new InvalidOperationException("Clarification already responded");

        Response = response;
        RespondedAt = DateTime.UtcNow;
        Status = ClarificationStatus.Responded;

        AddDomainEvent(new ClarificationRespondedEvent(Id, EmployeeId, RequestedById));
    }

    public void AddAttachment(ClarificationAttachment attachment)
    {
        _attachments.Add(attachment);
    }

    public void Close()
    {
        Status = ClarificationStatus.Closed;
    }

    public bool IsOverdue => DueDate.HasValue && DateTime.UtcNow > DueDate && Status == ClarificationStatus.Pending;
}

public class ClarificationAttachment : BaseEntity
{
    public Guid ClarificationId { get; private set; }
    public string FileName { get; private set; } = string.Empty;
    public string FileUrl { get; private set; } = string.Empty;
    public string FileType { get; private set; } = string.Empty;
    public long FileSize { get; private set; }

    private ClarificationAttachment() { }

    public static ClarificationAttachment Create(
        Guid clarificationId,
        string fileName,
        string fileUrl,
        string fileType,
        long fileSize)
    {
        return new ClarificationAttachment
        {
            ClarificationId = clarificationId,
            FileName = fileName,
            FileUrl = fileUrl,
            FileType = fileType,
            FileSize = fileSize
        };
    }
}

// Domain Events
public class ClarificationRequestedEvent : DomainEvent
{
    public Guid ClarificationId { get; }
    public Guid EmployeeId { get; }
    public Guid RequestedById { get; }
    public DateOnly StartDate { get; }
    public DateOnly EndDate { get; }

    public ClarificationRequestedEvent(
        Guid clarificationId,
        Guid employeeId,
        Guid requestedById,
        DateOnly startDate,
        DateOnly endDate)
    {
        ClarificationId = clarificationId;
        EmployeeId = employeeId;
        RequestedById = requestedById;
        StartDate = startDate;
        EndDate = endDate;
    }
}

public class ClarificationRespondedEvent : DomainEvent
{
    public Guid ClarificationId { get; }
    public Guid EmployeeId { get; }
    public Guid RequestedById { get; }

    public ClarificationRespondedEvent(Guid clarificationId, Guid employeeId, Guid requestedById)
    {
        ClarificationId = clarificationId;
        EmployeeId = employeeId;
        RequestedById = requestedById;
    }
}

File: attendance-system/mobile-app/lib/domain/entities/clarification.dart
import 'package:equatable/equatable.dart';

/// Clarification status enum
enum ClarificationStatus {
  pending,
  responded,
  closed,
}

/// Clarification request entity (sent by admin to employee)
class Clarification extends Equatable {
  final String id;
  final String employeeId;
  final String requestedBy;
  final String requestedByName;
  final DateTime startDate;
  final DateTime endDate;
  final String question;
  final ClarificationStatus status;
  final String? response;
  final List<ClarificationAttachment> responseAttachments;
  final DateTime? respondedAt;
  final DateTime createdAt;
  final DateTime? dueDate;

  const Clarification({
    required this.id,
    required this.employeeId,
    required this.requestedBy,
    required this.requestedByName,
    required this.startDate,
    required this.endDate,
    required this.question,
    required this.status,
    this.response,
    required this.responseAttachments,
    this.respondedAt,
    required this.createdAt,
    this.dueDate,
  });

  @override
  List<Object?> get props => [
        id,
        employeeId,
        requestedBy,
        requestedByName,
        startDate,
        endDate,
        question,
        status,
        response,
        responseAttachments,
        respondedAt,
        createdAt,
        dueDate,
      ];

  bool get isPending => status == ClarificationStatus.pending;
  bool get isResponded => status == ClarificationStatus.responded;
  bool get isClosed => status == ClarificationStatus.closed;
  bool get isOverdue => dueDate != null && DateTime.now().isAfter(dueDate!) && isPending;
}

/// Clarification attachment
class ClarificationAttachment extends Equatable {
  final String id;
  final String fileName;
  final String fileUrl;
  final String fileType;
  final int fileSize;

  const ClarificationAttachment({
    required this.id,
    required this.fileName,
    required this.fileUrl,
    required this.fileType,
    required this.fileSize,
  });

  @override
  List<Object?> get props => [id, fileName, fileUrl, fileType, fileSize];
}

/// Clarification response request
class ClarificationResponse extends Equatable {
  final String clarificationId;
  final String response;
  final List<String> filePaths;

  const ClarificationResponse({
    required this.clarificationId,
    required this.response,
    required this.filePaths,
  });

  @override
  List<Object?> get props => [clarificationId, response, filePaths];
}
